internal/storage/metastorage/firestore/event_storage.go

```diff
@@ -42,47 +55,403 @@ func newEventStorage(params Params, client *firestore.Client) (internal.EventSto
+	if err := e.validateEventTag(eventTag); err != nil {
+		return 0, err
+	}
+	return e.instrumentGetFirstEventIdByBlockHeight.Instrument(ctx, func(ctx context.Context) (int64, error) {
+		events, err := e.getEventsByBlockHeight(ctx, eventTag, blockHeight)
+		if err != nil {
+			return 0, xerrors.Errorf("failed to get events by block height: %w", err)
+		}
+		if len(events) == 0 {
+			return 0, errors.ErrItemNotFound
+		}
+		eventId := int64(math.MaxInt64)
+		for _, event := range events {
+			if event.EventId < eventId {
+				eventId = event.EventId
+			}
+		}
+		return eventId, nil
+	})
 }
 
 // GetMaxEventId implements internal.EventStorage.
-func (*eventStorageImpl) GetMaxEventId(ctx context.Context, eventTag uint32) (int64, error) {
-	panic("unimplemented")
+func (e *eventStorageImpl) GetMaxEventId(ctx context.Context, eventTag uint32) (int64, error) {
+	if err := e.validateEventTag(eventTag); err != nil {
+		return 0, err
+	}
+	return e.instrumentGetMaxEventId.Instrument(ctx, func(ctx context.Context) (int64, error) {
+		maxEventId, err := e.getMaxEventId(ctx, eventTag, nil)
+		if err != nil {
+			return 0, nil
+		}
+		if maxEventId == model.EventIdDeleted {
+			return 0, errors.ErrNoEventHistory
+		}
+		return maxEventId, nil
+	})
 }
 
 // SetMaxEventId implements internal.EventStorage.
-func (*eventStorageImpl) SetMaxEventId(ctx context.Context, eventTag uint32, maxEventId int64) error {
-	panic("unimplemented")
+func (e *eventStorageImpl) SetMaxEventId(ctx context.Context, eventTag uint32, maxEventId int64) error {
+	if err := e.validateEventTag(eventTag); err != nil {
+		return err
+	}
+	if maxEventId < model.EventIdStartValue && maxEventId != model.EventIdDeleted {
+		return xerrors.Errorf("invalid max event id: %d", maxEventId)
+	}
+	return e.instrumentSetMaxEventId.Instrument(ctx, func(ctx context.Context) error {
+		return e.client.RunTransaction(ctx, func(ctx context.Context, t *firestore.Transaction) error {
+			currentMaxEventId, err := e.getMaxEventId(ctx, eventTag, t)
+			if err != nil {
+				return xerrors.Errorf("failed to get max event id: %w", err)
+			}
+			if maxEventId > currentMaxEventId {
+				return xerrors.Errorf("can not set max event id to be %d, which is bigger than current max event id: %d", maxEventId, currentMaxEventId)
+			}
+			if maxEventId != model.EventIdDeleted {
+				newLatestEventDocRef := e.getEventDocRef(eventTag, maxEventId)
+				newLatestEventDoc, err := t.Get(newLatestEventDocRef)
+				if err != nil {
+					return xerrors.Errorf("failed to get event entry with new max event id: %d, error: %w", maxEventId, err)
+				}
+				if !newLatestEventDoc.Exists() {
+					return xerrors.Errorf("event entry with new max event id %d does not exist", maxEventId)
+				}
+			}
+			return e.setMaxEventId(ctx, eventTag, maxEventId, t)
+		})
+	})
+}
+
+/////////////////////////////////////////////////////////
+// firestore storage
+
+// Fields in the payload map are configured not to be automatically indexed.
```

Comment:
Since we're configuring the index for exclusions, can we just use a flat structure, rather than a nested struct?

Reply:
This is because there is a limit of maximum number of single field configurations across the database: https://cloud.google.com/firestore/quotas#indexes

We only use 1 configuration for nested struct, subfields inherit from it.